File: solutions/haskell/book-store/1/src/BookStore.hs
-- respectfully dedicated to Shree DR.MDD
module BookStore (total, Book(..)) where

import Data.List (sort, group)

data Book = First | Second | Third | Fourth | Fifth deriving (Eq, Ord, Show)

total :: [Book] -> Int
total items = compute 5 $ reverse $ sort $ map length $ group $ sort items

compute :: Int -> [Int] -> Int
compute _ [] = 0
compute 1 xs = sum $ map (\v -> groupCost 1 * v) xs
compute k xs
    | length xs < k = compute (k - 1) xs
    | otherwise     = min (groupCost k + compute k updated) (compute (k - 1) xs)
  where
    updated = filter (/= 0) (map (subtract 1) $ take k xs) ++ drop k xs

groupCost :: Int -> Int
groupCost 1 =  800
groupCost 2 = 1520
groupCost 3 = 2160
groupCost 4 = 2560
groupCost 5 = 3000
groupCost _ =    0

File: solutions/php/book-store/1/BookStore.php
<?php
// Dedicated to Shree DR.MDD

function total(array $items): int
{
    $buckets = array_fill(1, 5, 0);
    $collections = array_fill(1, 5, 0);

    foreach ($items as $id) {
        $buckets[$id]++;
    }

    while (array_sum($buckets) > 0) {
        $currentSet = 0;
        foreach ($buckets as $key => $count) {
            if ($count > 0) {
                $currentSet++;
                $buckets[$key]--;
            }
        }
        if ($currentSet > 0) {
            $collections[$currentSet]++;
        }
    }

    while ($collections[3] > 0 && $collections[5] > 0) {
        $collections[3]--;
        $collections[5]--;
        $collections[4] += 2;
    }

    $totalCost = 800 * $collections[1] +
        (800 * 2 * 0.95) * $collections[2] +
        (800 * 3 * 0.9) * $collections[3] +
        (800 * 4 * 0.8) * $collections[4] +
        (800 * 5 * 0.75) * $collections[5];

    return (int)$totalCost;
}
